### PR TITLE
Queen Sacrifice Filter

### PR DESCRIPTION
**Pull request by josephwilliams**
Partial resolution of issue https://github.com/agadmator-library/agadmator-library.github.io/issues/86

This adds a "Queen Sacrifice" filter, selectable by checkbox in the UI after the underpromotion filter.

The logic follows black and white queen positions during games. When a queen is captured, it checks if the capturing piece was a queen. If so, this would be an exchange, not a sacrifice. It then checks if the next move was a capture of the other queen, which would also denote an exchange. In the case that a queen is captured, and the opposing queen is not recaptured in the next move, and it is not checkmate, nor does resignation occur, we consider this to be a queen sacrifice.

An exception: this does not take into consideration combinations in which there could be a series of moves before a queen is recaptured. It also does not take into consideration a pawn promoted to a queen that is thereafter sacrificed.

### CODE REVIEW

File: src/filter/QueenSacrificeFilter.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  There are also some cases where the queen sacrifice may not be detected:
+  - This function does not currently handle the scenario in which a queen exchange takes place and, later, a pawn is promoted to a queen, and then sacrificed.
+*/
+function isQueenSacrifice(pgn: string): boolean {
+  const moves = pgn
+    .replace(/\d+\.\s+/g, "") // Remove move numbers
+    .trim()
+    .split(/\s+/); // Split by spaces
+
+  const captureRegex = /([QRNB])?([a-h])?x([a-h][1-8])?/;
```
Comment (agadmator-library) [inline src/filter/QueenSacrificeFilter.ts:20]: This regex misses situation where two kinghts are located in single file. In that case notation is `N3xa5`

File: src/filter/QueenSacrificeFilter.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        // If the next move is not a queen recapture, and it’s not an exchange, return true
+        return !nextMoveIsQueenCapture && i !== lastMoveIndex;
+      }
+    }
+
+    // Update queen positions if the queens move
+    if (move.startsWith("Q")) {
+      if (i % 2 === 0) {
+        // White's turn (even index)
+        const newSquare = move.slice(1); // Extract the square from the move (e.g., Qd2 -> d2)
```
Comment (agadmator-library) [inline src/filter/QueenSacrificeFilter.ts:74]: I think this fails to detect queen move if queen is performing capture `Qxd3` becomes `xd3`

File: index.html [hunk 3]
```diff
@@ -35,7 +35,7 @@
 <body>
 
 <div id="app"></div>
-<script src="generated/js/references-00ad6ccb4d2b9092d23043deedfe0aa81cfe64fe.js"></script>
+<script src="generated/js/references-dd28231060fd294b989f08ab819540d2dce563ad.js"></script>
```
Comment (agadmator-library) [inline index.html:38]: Please exclude this change from PR as this section is autogenerated

File: src/filter/QueenSacrificeFilter.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    // If it's a capture move
+    if (captureMatch) {
+      const targetSquare = captureMatch[3]; // The square where the capture happens
+
+      const isWhiteQueenCapture = targetSquare === whiteQueenPosition;
+      const isBlackQueenCapture = targetSquare === blackQueenPosition;
+
+      if (isWhiteQueenCapture) {
+        if (captureMatch[1] === "Q") {
+          // If it's a queen capturing a queen, it's an exchange, not a sacrifice
```
Comment (agadmator-library) [inline src/filter/QueenSacrificeFilter.ts:41]: Actually it is not. It is only exchange if there will be immediate exchange on same field.

Consider:
`20. Qxd5 /Black queen capture/ h6 /random pawn move by black/ 21. Qd5 /White queen is stil in game/`

To handle this situation you would have to check if next move is capture on same field.

**Review by agadmator-library — CHANGES_REQUESTED** [review]
Hi, thanks for this filter. I think checking only next move can be good approximation. I added some comments

File: src/filter/QueenSacrificeFilter.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    const captureMatch = move.match(captureRegex);
+
+    // If it's a capture move
+    if (captureMatch) {
+      const targetSquare = captureMatch[3]; // The square where the capture happens
+
+      const isWhiteQueenCapture = targetSquare === whiteQueenPosition;
+      const isBlackQueenCapture = targetSquare === blackQueenPosition;
+
+      if (isWhiteQueenCapture) {
```
Comment (agadmator-library) [inline src/filter/QueenSacrificeFilter.ts:39]: I think you can easliy get rid of duplication in conditional branches by doing something like this

```
if (isWhiteQueenCapture || isBlackQueenCapture) {
  var otherQueenPosition = isWhiteQueenCapture ? blackQueenPosition : whiteQueenPosition
  if (captureMatch[i] === "Q") {
    otherQueenPosition = targetSquare
  }
  const nextMove = moves[i + 1];
  const nextCaptureMatch = nextMove?.match(captureRegex);
  const nextTargetSquare = nextCaptureMatch?.[3];
  const nextMoveIsQueenCapture = nextTargetSquare === otherQueenPosition;
  return !nextMoveIsQueenCapture && i !== lastMoveIndex;
}
```

File: src/filter/QueenSacrificeFilter.ts (hunk 2 above)
Comment (josephwilliams) [reply src/filter/QueenSacrificeFilter.ts:74]: Fixed for this, and also to handle moves postfixed with "-" for check and "#" for mate.

File: index.html (hunk 3 above)
Comment (josephwilliams) [reply index.html]: Excluded.

File: src/filter/QueenSacrificeFilter.ts (hunk 4 above)
Comment (josephwilliams) [reply src/filter/QueenSacrificeFilter.ts]: True, and re-capture of queen on next move check is part of original logic.

File: src/filter/QueenSacrificeFilter.ts (hunk 1 above)
Comment (josephwilliams) [reply src/filter/QueenSacrificeFilter.ts]: Fixed to handle disambiguated moves.

**Comment by josephwilliams** [comment]
Came across a mismatched PGN while testing: https://github.com/agadmator-library/agadmator-library.github.io/issues/153

**Review by agadmator-library — APPROVED** [review]

**Comment by agadmator-library** [comment]
Thanks for contribution. Lets see it in action :)

**Pull request merged**